lib/IRGen/GenCall.cpp

```diff
@@ -1960,24 +1969,24 @@ bool irgen::addNativeArgument(IRGenFunction &IGF, Explosion &in,
 }
 
 /// Emit a direct parameter that was passed under a C-based CC.
-static void emitDirectForeignParameter(IRGenFunction &IGF,
-                                       Explosion &in,
-                                       llvm::Type *coercionTy,
-                                       Explosion &out,
-                                       SILType paramType,
+static void emitDirectForeignParameter(IRGenFunction &IGF, Explosion &in,
+                                       const clang::CodeGen::ABIArgInfo &AI,
+                                       Explosion &out, SILType paramType,
                                        const LoadableTypeInfo &paramTI) {
   // The ABI IR types for the entrypoint might differ from the
   // Swift IR types for the body of the function.
 
-  ArrayRef<llvm::Type*> expandedTys;
-  if (auto expansionTy = dyn_cast<llvm::StructType>(coercionTy)) {
-    expandedTys = makeArrayRef(expansionTy->element_begin(),
-                               expansionTy->getNumElements());
+  llvm::Type *coercionTy = AI.getCoerceToType();
 
-  // Fast-path a really common case.  This check assumes that either
-  // the storage type of a type is an llvm::StructType or it has a
-  // single-element explosion.
+  ArrayRef<llvm::Type*> expandedTys;
+  if (AI.isDirect() && AI.getCanBeFlattened() &&
```

Comment:
The same isDirectFlattened || isIndirect logic should apply here too. We are also calling this for indirect parameters. Or am I mistaken?

This should make any changes in the existing tests unneccessary. I believe with the change as is lowering of indirect (eg. i8*) parameters at function entry still goes through memory. The optimizer would optimize this away but at Onone this is unecessary overhead.

Reply:
Isn't that incorrect though?  I thought that we don't want to flatten the parameter unless we are passing it directly.

Reply:
Oh sorry, I was getting confused here this is only checking for the case where we have to expand. An indirect type would also never be an llvm::StructType.

You are right of course.